common/remote/hooks/git/gitfile.cpp

```diff
@@ -410,10 +413,329 @@ class GitRepositoryDirectoryIterator : implements IDirectoryIterator, public CIn
     }
 };
 
+
+//--------------------------------------------------------------------------------------------------------------------
+// New implementation using libgit2
+//--------------------------------------------------------------------------------------------------------------------
+
+#define GIT_CHECK(x) check(x, #x)
+class GitCommitTree : public CInterface
+{
+public:
+    GitCommitTree(const char * directory, const char * version)
+    {
+        GIT_CHECK(git_repository_open(&gitRepo, directory));
+
+        if (gitRepo)
+        {
+            //Check to see if the version is a tag/branch etc. - these take precedence if they happen to match a sha prefix
+            git_reference * ref = nullptr;
+            if (git_reference_dwim(&ref, gitRepo, version) == 0)
+            {
+                //Map the symbolic reference to the underlying object
+                git_reference * resolvedRef = nullptr;
+                if (git_reference_resolve(&resolvedRef, ref) == 0)
+                {
+                    const git_oid * oid = git_reference_target(resolvedRef);
+                    check(git_commit_lookup(&gitCommit, gitRepo, oid), "git_commit_lookup");
+                    git_reference_free(resolvedRef);
+                }
+
+                git_reference_free(ref);
+            }
+
+            if (!gitCommit)
+            {
+                git_oid gitOid;
+                if (git_oid_fromstrp(&gitOid, version) == 0)
+                {
+                    //User provided a SHA (possibly shorted) -> resolve it.  Error will be reported later if it does not exist.
+                    check(git_commit_lookup_prefix(&gitCommit, gitRepo, &gitOid, strlen(version)), "git_commit_lookup_prefix");
+                }
+            }
+
+            if (gitCommit)
+                check(git_commit_tree(&gitRoot, gitCommit), "git_commit_tree");
+        }
+    }
+    ~GitCommitTree()
+    {
+        git_tree_free(gitRoot);
+        git_commit_free(gitCommit);
+        git_repository_free(gitRepo);
+    }
+
+    const git_tree * queryTree() const { return gitRoot; }
+
+protected:
+    void check(int code, const char * func)
+    {
+        if (code != 0)
+        {
+            const git_error * err = git_error_last();
+            const char * errmsg = err ? err->message : "<unknown>";
+            WARNLOG("libgit %s returned %u: %s", func, code, errmsg);
+        }
+    }
+protected:
+    git_repository * gitRepo = nullptr;
+    git_commit * gitCommit = nullptr;
+    git_tree * gitRoot = nullptr;
+};
+
+
+class NewGitRepositoryFileIO : implements CSimpleInterfaceOf<IFileIO>
+{
+public:
+    NewGitRepositoryFileIO(const char * gitDirectory, const char * revision, const char * relFileName, GitCommitTree * commitTree, const git_oid * oid)
```

Comment:
First three parameters appear to be unused

Reply:
removed